interface/src/app/gene/search.js

```diff
@@ -0,0 +1,402 @@
+            return false;
+          }
+          $scope.loadingSearchResults = true;
+
+          $rootScope.$broadcast('results.loadingSearchResults');
+
+          var qparams = {
+            'query': $scope.genesToAdd.query
+          };
+
+          if ($scope.organism) {
+            // TODO: adage doesn't currently support multiple organisms.
+            // Therefore, when retrieving gene objects from the API gene
+            // search endpoint, this will return *all* gene objects it finds,
+            // without filtering for organism. At the point when we add
+            // genes for more than one species to the database, we will
+            // need to set this $scope.organism to whatever organism is in
+            // the ML model (or genes for many organisms will be returned).
+            qparams['organism'] = $scope.organism;
+          }
+
+          SearchResults.search(qparams);
+        };
+
+        $scope.$on('results.searchResultsReturned', function() {
+          $scope.loadingSearchResults = false;
+        });
+      }
+    ],
+    replace: true,
+    restrict: "E",
+    scope: {
+      query: '@'
+    },
+    templateUrl: 'gene/gene-search-form.tpl.html'
+  };
+}])
+
+// Directive for table containing search results
+.directive('searchResultTable', ['SearchResults', function(SearchResults) {
+  return {
+    controller: ['$scope', 'SearchResults', 'SelectedGenesFactory',
+      function($scope, SearchResults, SelectedGenesFactory) {
+        $scope.currentPage = 1;
+        $scope.itemsPerPage = 10;
+        $scope.totalResults = 0;
+        $scope.maxSize = 10;
+        $scope.resultsForPage = [];
+        $scope.searchResults = SearchResults.getSearchResults();
+        $scope.loadingSearchResults = false;
+
+        $scope.addAllNonAmbiguous = function() {
+          // Function to automatically add all genes that
+          // only have one search result.
+          var searchResults = SearchResults.getSearchResults();
+          for (var key in searchResults) {
+            var results = searchResults[key];
+            if (results.found.length <= 1) {
+              if (results.found[0]) {
+                SelectedGenesFactory.addGene(results.found[0]);
+                SearchResults.remove(key);
+              }
+            }
+          }
+        };
+
+        $scope.removeNotFound = function() {
+          // Function to get rid of all the queries that returned no results.
+          var searchResults = SearchResults.getSearchResults();
+          for (var key in searchResults) {
+            var results = searchResults[key];
+            if (results.found.length === 0) {
+              SearchResults.remove(key);
+            }
+          }
+        };
+      }
+    ],
+
+    link: function(scope, element, attr) {
+      scope.$on('results.update', function() {
+        scope.totalResults = SearchResults.size();
+        var begin = ((scope.currentPage - 1) * scope.itemsPerPage);
```

Comment:
Not sure whether you saw my previous comments. Lines 203-205 are identical to 218-220. Not sure whether it is worth the time to factor them out.